File: branches/version-1.8/IPViewer/IPViewerSettings.cpp
/*
	Copyright 2011 Justin LeCheminant

	This file is part of IP Viewer.

    filesort is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    filesort is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with filesort.  If not, see <http://www.gnu.org/licenses/>.
*/

// IPViewerSettings.cpp : implementation file
//

#include "stdafx.h"
#include "IPViewer.h"
#include "IPViewerSettings.h"
#include "IPSettings.h"

// CIPViewerSettings dialog

IMPLEMENT_DYNAMIC(CIPViewerSettings, CDialog)

CIPViewerSettings::CIPViewerSettings(CWnd* pParent /*=NULL*/)
	: CDialog(CIPViewerSettings::IDD, pParent)
	, m_bMinToTray(FALSE)
	, m_bCloseTray(FALSE)
	, m_nTick(0)
	, m_nSelectedIndex(0)
	, m_bOnTop(FALSE)
	, m_bStartMinimized(FALSE)
	, m_bCheckExternalIp(FALSE)
{
	m_hIcon = AfxGetApp()->LoadIcon(IDI_ICON_IPVIEWER);
}

CIPViewerSettings::CIPViewerSettings( CWnd* pParent /*=NULL*/, CIPSettings* settings)
	: CDialog( CIPViewerSettings::IDD, pParent)
{
	m_hIcon = AfxGetApp()->LoadIcon(IDI_ICON_IPVIEWER);
	m_pSettings = settings;

}

CIPViewerSettings::~CIPViewerSettings()
{
}

void CIPViewerSettings::DoDataExchange(CDataExchange* pDX)
{
	CDialog::DoDataExchange(pDX);
	DDX_Check(pDX, IDC_CHECK_MINTRAY, m_bMinToTray);
	DDX_Check(pDX, IDC_CHECK_CLOSETRAY, m_bCloseTray);
	DDX_Check(pDX, IDC_CHECK1, m_bOnTop);
	DDX_Text(pDX, IDC_EDIT_RATE, m_nTick);
	DDX_Radio(pDX, IDC_RADIO_IP, m_nSelectedIndex);	
	DDX_Check(pDX, IDC_CHECK2, m_bStartMinimized);
	DDX_Check(pDX, IDC_CHECK_EXTERNAL, m_bCheckExternalIp);
}


BOOL CIPViewerSettings::OnInitDialog()
{
	CDialog::OnInitDialog();

	// TODO: init dialog here
	SetIcon(m_hIcon, TRUE);			// Set big icon
	SetIcon(m_hIcon, FALSE);		// Set small icon

	// load our dialog data
	m_bMinToTray = m_pSettings->GetMinimizeToTray();
	m_bCloseTray = m_pSettings->GetCloseToTray();
	m_nTick = m_pSettings->GetTick();
	m_nSelectedIndex = m_pSettings->GetShowItemIndex();
	m_bOnTop = m_pSettings->GetOnTop();
	m_bStartMinimized = m_pSettings->GetStartMinimized();
	m_bCheckExternalIp = m_pSettings->GetCheckExternalIp();

	// refresh the Dialog
	UpdateData( FALSE );

	return TRUE;

}

BEGIN_MESSAGE_MAP(CIPViewerSettings, CDialog)
	ON_BN_CLICKED(IDC_BUTTON_SAVE, &CIPViewerSettings::OnBnClickedButtonSave)
	ON_BN_CLICKED(IDC_BUTTON_CANCEL, &CIPViewerSettings::OnBnClickedButtonCancel)
END_MESSAGE_MAP()


// CIPViewerSettings message handlers

void CIPViewerSettings::OnBnClickedButtonSave()
{
	// turn on wait cursor
	CDialog::BeginWaitCursor();
	
	// get data back from the dialog
	UpdateData( TRUE );

	// set our values
	m_pSettings->SetMinimizeToTray( m_bMinToTray );
	m_pSettings->SetCloseToTray( m_bCloseTray );
	m_pSettings->SetTick( m_nTick );
	m_pSettings->SetShowItemIndex( m_nSelectedIndex );
	m_pSettings->SetOnTOp( m_bOnTop );
	m_pSettings->SetStartInTray( m_bStartMinimized );
	m_pSettings->SetCheckExternalIp( m_bCheckExternalIp );

	// save our changes to the registry
	CString result = m_pSettings->Save() ? 
		_T("Settings were saved successfully.") : 
		_T("An error occured saving settings.");

	AfxMessageBox( result, MB_OKCANCEL );
		
	// turn on normal cursor
	CDialog::EndWaitCursor();
	
	// then kill the dialog
	this->EndDialog(0);
}

void CIPViewerSettings::OnBnClickedButtonCancel()
{
	CDialog::EndDialog(0);

	// kill this dialog
	this->EndDialog(0);
}

File: IpViewer.BusinessObject/IpViewerSettings.cs
//   Copyright(C) 2020 Justin LECheminant


//   This program is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with this program.If not, see<https://www.gnu.org/licenses/>.

using System.Configuration;

namespace IpViewer.BusinessObject
{
    public class IpViewerSettings
    {
        #region

        public static IpViewerSettings _instance;

        public static object _lock = new object();

        #endregion

        public static IpViewerSettings Instance
        {
            get
            {
                lock (_lock)
                {
                    if (_instance == null)
                    {
                        _instance = new IpViewerSettings();
                        _instance.LoadSettings();

                    }
                }

                return _instance;
            }
        }

        private IpViewerSettings()
        {

        }

        #region Properties

        public bool Use12HourClock { get; set; }

        public bool GetExternalAddress { get; set; }

        public bool StartMinimized { get; set; }

        public bool LogExternalIpAddress { get; set; }

        public bool TopMost { get; set; }

        public bool CloseToTray { get; set; }

        public int RefreshTime { get; set; }

        public bool MinimizeToTray { get; set; }

        public bool LogAdapterOperationalStatus { get; set; }

        public string CurrentAdapterName { get; set; }

        public bool LoggingEnabled { get; set; }

        public bool LogAdapterInformation { get; set; }

        public bool LogAdapterIpAddress { get; set; }
        
        public bool LogIpAddress { get; set; }

        public double Opacity { get; set; }

        #endregion

        private void LoadSettings()
        {
            RefreshTime = int.Parse(ConfigurationManager.AppSettings["RefreshTime"]);
            Opacity = double.Parse(ConfigurationManager.AppSettings["Opacity"]);
            LoggingEnabled = bool.Parse(ConfigurationManager.AppSettings["LoggingEnabled"]);
            LogIpAddress = bool.Parse(ConfigurationManager.AppSettings["LogIpAddress"]);
            MinimizeToTray = bool.Parse(ConfigurationManager.AppSettings["MinimizeToTray"]);
            CloseToTray = bool.Parse(ConfigurationManager.AppSettings["CloseToTray"]);
            StartMinimized = bool.Parse(ConfigurationManager.AppSettings["StartMinimized"]);
            GetExternalAddress = bool.Parse(ConfigurationManager.AppSettings["GetExternalAddress"]);
            TopMost = bool.Parse(ConfigurationManager.AppSettings["Topmost"]);
            CurrentAdapterName = ConfigurationManager.AppSettings["CurrentAdapterName"];
            LogAdapterInformation = bool.Parse(ConfigurationManager.AppSettings["LogAdapterInformation"]);
            LogAdapterIpAddress = bool.Parse(ConfigurationManager.AppSettings["LogAdapterIpAddress"]);
            LogAdapterOperationalStatus = bool.Parse(ConfigurationManager.AppSettings["LogAdapterOperationalStatus"]);
            Use12HourClock = bool.Parse(ConfigurationManager.AppSettings["Use12HourClock"]);
        }

        public void SaveSettings()
        {
            var manager = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None);

            manager.AppSettings.Settings["RefreshTime"].Value = RefreshTime.ToString();
            manager.AppSettings.Settings["Opacity"].Value = Opacity.ToString();
            manager.AppSettings.Settings["LoggingEnabled"].Value = LoggingEnabled.ToString();
            manager.AppSettings.Settings["LogIpAddress"].Value = LogIpAddress.ToString();
            manager.AppSettings.Settings["MinimizeToTray"].Value = MinimizeToTray.ToString();
            manager.AppSettings.Settings["CloseToTray"].Value = CloseToTray.ToString();
            manager.AppSettings.Settings["StartMinimized"].Value = StartMinimized.ToString();
            manager.AppSettings.Settings["GetExternalAddress"].Value = GetExternalAddress.ToString();
            manager.AppSettings.Settings["Topmost"].Value = TopMost.ToString();
            manager.AppSettings.Settings["CurrentAdapterName"].Value = CurrentAdapterName;
            manager.AppSettings.Settings["LogAdapterInformation"].Value = LogAdapterInformation.ToString();
            manager.AppSettings.Settings["LogAdapterIpAddress"].Value = LogAdapterIpAddress.ToString();
            manager.AppSettings.Settings["LogAdapterOperationalStatus"].Value = LogAdapterOperationalStatus.ToString();
            manager.AppSettings.Settings["Use12HourClock"].Value = Use12HourClock.ToString();

            manager.Save(ConfigurationSaveMode.Full, true);
            ConfigurationManager.RefreshSection("appSettings");
        }

    }
}